Documentation/Evolution/StringProcessingAlgorithms.md

```diff
@@ -0,0 +1,544 @@
+    /// around elements equal to the given separator.
+    /// - Parameter separator: A regex describing elements to be split upon.
+    /// - Returns: A collection of substrings, split from this collection's
+    /// elements.
+    public func split<R: RegexComponent>(by separator: R) -> some Collection<Substring>
+}
+```
+
+
+### `CustomMatchingRegexComponent`
+
+```swift
+/// A protocol for custom match functionality.
+public protocol CustomMatchingRegexComponent : RegexComponent {
+    /// Match the input string within the specified bounds, beginning at the given index, and return
+    /// the end position (upper bound) of the match and the matched instance.
+    /// - Parameters:
+    ///   - input: The string in which the match is performed.
+    ///   - index: An index of `input` at which to begin matching.
+    ///   - bounds: The bounds in `input` in which the match is performed.
+    /// - Returns: The upper bound where the match terminates and a matched instance, or `nil` if
+    ///   there isn't a match.
+    func match(
+        _ input: String,
+        startingAt index: String.Index,
+        in bounds: Range<String.Index>
+    ) -> (upperBound: String.Index, match: Match)?
+}
+```
+
+`CustomMatchingRegexComponent` inherits from `RegexComponent` and satisfies its sole requirement; Conformers can be used with all of the string algorithms generic over `RegexComponent`.
+
+Here, we use Foundation `FloatingPointFormatStyle<Double>.Currency` as an example for protocol conformance. It would implement the `match` function with `Match` being a `Double`. It could also add a static function `.localizedCurrency(code:)` as a member of `RegexComponent`, so it can be referred as `.localizedCurrency(code:)` in the `Regex` result builder:
+
+```swift
+extension FloatingPointFormatStyle<Double>.Currency : CustomMatchingRegexComponent { 
+    public func match(
+        _ input: String,
+        startingAt index: String.Index,
+        in bounds: Range<String.Index>
+    ) -> (upperBound: String.Index, match: Double)?
+}
+
+extension RegexComponent where Self == FloatingPointFormatStyle<Double>.Currency {
+    public static func localizedCurrency(code: Locale.Currency) -> Self
+}
+```
+
+Matching and extracting a localized currency amount, such as `"$3,020.85"`, can be done directly within a regex:
+
+```swift
+let regex = Regex {
+    capture(.localizedCurreny(code: "USD"))
+}
+```
+
+
+## Alternatives considered
+
+### Extend `Sequence` instead of `Collection`
+
+All of the proposed algorithms are specific to the `Collection` protocol, without support for plain `Sequence`s. Types conforming to the `Sequence` protocol are not required to support multi-pass iteration, which makes a `Sequence` conformance insufficient for most of these algorithms. In light of this, the decision was made to have the underlying shared algorithm implementations work exclusively with `Collection`s.
```

Comment:
```suggestion
Most of the proposed algorithms are necessarily on `Collection` due to the use of indices or mutation.  `Sequence` does not support multi-pass iteration, so even `trimPrefix` would problematic on `Sequence` because it needs to look 1 `Element` ahead to know when to stop trimming.
```